container: Introduce chunked_hash_map

StephanDollberg opened the pull request:
Introduces `chunked_hash_map` which is a type alias for the segmented
version of https://github.com/martinus/unordered_dense/ using our
`chunked_vector` as a backing array.

This allows us to use a hashmap in scenarios where the map may grow
large (i.e.: scale with partition/topic counts) similar to `chunked_vector`
and avoids us having to fall back to the suboptimal `absl::btree_map`.

Below are some numbers from our leader balanacer benchmark comparing the
different map types:

| test                                                                  |  iterations |      median |         mad |         min |         max |      allocs |       tasks |        inst |
| -                                                                     |           - |           - |           - |           - |           - |           - |           - |           - |
| lb.random_generator_movement_absl_node                                |         149 |     3.120ms |    97.624ns |     3.120ms |     3.122ms |       0.000 |       0.000 |  44663355.3 |
| lb.random_generator_movement_absl_flat                                |         152 |     3.090ms |   342.546ns |     3.089ms |     3.090ms |       0.000 |       0.000 |  44846754.1 |
| lb.random_generator_movement_absl_btree                               |          84 |     8.103ms |   884.595ns |     8.102ms |     8.105ms |       0.000 |       0.000 |  57163339.9 |
| lb.random_generator_movement_std_map                                  |         117 |     5.028ms |   676.256ns |     5.027ms |     5.030ms |       0.000 |       0.000 |  53678485.9 |
| lb.random_generator_movement_unordered_dense                          |         156 |     2.850ms |   576.237ns |     2.849ms |     2.851ms |       0.000 |       0.000 |  43115077.3 |
| lb.random_generator_movement_unordered_dense_segmented                |         149 |     3.111ms |   360.074ns |     3.111ms |     3.115ms |       0.000 |       0.000 |  46250454.0 |
| lb.random_generator_movement_unordered_dense_segmented_frag           |         151 |     3.063ms |   388.795ns |     3.063ms |     3.065ms |       0.000 |       0.000 |  45697581.1 |
| lb.random_generator_movement_unordered_dense_segmented_frag_absl_hash |         139 |     3.643ms |   444.705ns |     3.642ms |     3.643ms |       0.000 |       0.000 |  46476708.7 |


## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.3.x
- [ ] v23.2.x

## Release Notes

* none

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/46439#018e5760-050a-47b9-a2ab-18438a5adc64:
```
"rptest.tests.recovery_mode_test.DisablingPartitionsTest.test_disable"
```
<!--197a962e-7769-4630-bde2-d399d3a65fa5-->

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/46439#018e5760-050a-47b9-a2ab-18438a5adc64
<!--757f14cc-f911-45f1-b87e-19907283c9eb-->

rockwotj commented on src/v/container/include/container/fragmented_hash_map.h:
```diff
@@ -0,0 +1,66 @@
+    ankerl::unordered_dense::hash<Key>>,
+  typename EqualTo = std::equal_to<Key>>
+using fragmented_hash_map = ankerl::unordered_dense::
+  map<Key, Value, Hash, EqualTo, fragmented_vector<std::pair<Key, Value>>>;
```
Should we use `chunked_vector` to get the resize improvements (it's not on by default yet)

StephanDollberg replied:
Oh absolutely, this is a good catch. Using chunked_vector was my intention. I had thought we made dynamic_extent the default already.

I will change that tomorrow together with the name change.

StephanDollberg commented on cmake/dependencies.cmake:
```diff
@@ -53,6 +53,11 @@ fetch_dep(rapidjson
 
+fetch_dep(unordered_dense
+  REPO https://github.com/redpanda-data/unordered_dense
+  TAG redpanda
```
I went with this as that is what we are doing for some of our other forks (e.g.: seastar, croaring) but we are not consistent. I am happy either way.

dotnwat replied:
We should definitely use the sha1 in 3rdparty.cmake @ vtools. Here in the open-source build dependencies are more loose. But they probably shouldn't be. I'll make it a point to go add sha1's here as well.

StephanDollberg replied:
In vtools it's the sha. I will change the OSS one to use the sha as well then if we want to change the other ones as well.

StephanDollberg commented on src/v/container/include/container/fragmented_hash_map.h:
```diff
@@ -0,0 +1,66 @@
+ * Both std::hash and abseil's AbslHashValue are supported. We dispatch to the
+ * later if available. Given AbslHashValue also supports std::hash we could also
+ * unconditionally dispatch to it. However, it's hash mixing seems more
+ * extensive (and hence less performant) so we only do that when needed.
```
Meant absl hash (std hash doesn't do any hash mixing). Have clarified that.

StephanDollberg commented:
In https://github.com/redpanda-data/redpanda/compare/195bc196e1f959829746b9753d9e081b23fd6b92..46fd9b4762f743bbbde7ff7793f86858196c629b

 - Fixed typos
 - Renamed to chunked_hash_map
 - Actually use chunked_vector

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/46489#018e5b88-2d88-4ddf-bd09-5435f962240f:
```
"rptest.tests.simple_e2e_test.SimpleEndToEndTest.test_relaxed_acks.write_caching=False"
```
<!--bc7bdfa2-78b4-4cc6-a0d8-65cb775974ca-->

emaxerrno commented:
this is neat!

emaxerrno commented:
Given the intent is to use as widely as possible maybe the name of “fmap” is better. For commonly used classes, small tokens is better.

rockwotj commented on src/v/container/include/container/fragmented_vector.h:
```diff
@@ -70,10 +70,15 @@ class fragmented_vector {
     using const_reference = const T&;
     using size_type = size_t;
+    using allocator_type = backing_type::allocator_type;
+    using difference_type = backing_type::difference_type;
+    using pointer = backing_type::pointer;
+    using const_pointer = backing_type::const_pointer;
```
Why is this the difference/pointer type of `backing_type` and not `backing_type::value_type`?

StephanDollberg replied:
Yeah I guess it should be what you say. The allocator business always makes this complicated.

This however made me realize that I missed a thing in my patch to unordered_dense and we currently wouldn't use our custom type for the bucket array (but their segmented_vector) so I will need to fix that first.

StephanDollberg replied:
Fixed now

rockwotj reviewed (commented):
LGTM, a question and FYI the cover letter needs updating from the new name

StephanDollberg commented:
Failure is:
 - https://github.com/redpanda-data/redpanda/issues/16561

rockwotj commented on cmake/dependencies.cmake:
```diff
@@ -53,6 +53,12 @@ fetch_dep(rapidjson
+fetch_dep(unordered_dense
+  REPO https://github.com/redpanda-data/unordered_dense
+  TAG 9338f301522a965309ecec58ce61f54a52fb5c22
+  SOURCE_DIR redpanda_build
```
this directory doesn't exist in the repo copy paste error?

rockwotj replied:
Also you need to add this to MakeAvailable

StephanDollberg replied:
Ah brilliant, yeah I was still experimenting with this (== guessing).

rockwotj reviewed (commented):
lgtm, I think we need to fix the oss build

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/46639#018e6804-9e75-4332-89b5-60c36874181a
<!--d61c87cb-a162-4c7b-a979-c7503f0faf75-->

rockwotj commented on src/v/container/CMakeLists.txt:
```diff
@@ -1,7 +1,7 @@
 find_package(Roaring REQUIRED)
 v_cc_library(
   NAME container
-  DEPS Roaring::roaring
```
```
FAILED: lib/libv_v_io.so
: && /vectorized/llvm/bin/clang++ -fPIC -fPIC -march=westmere -fuse-ld=lld -Wno-unused-lambda-capture -Wno-unused-command-line-argument -stdlib=libc++ -fsanitize=undefined -fsanitize=address -fstack-protector-strong -fno-plt -fstack-clash-protection -O0 -g -gdwarf-4 -gsplit-dwarf  -Wl,--build-id=sha1 -L/vectorized/lib -L/vectorized/lib64 -stdlib=libc++ -fsanitize=undefined -fsanitize=address -Wl,-z,relro,-z,now -shared -Wl,-soname,libv_v_io.so -o lib/libv_v_io.so src/v/io/CMakeFiles/v_io.dir/persistence.cc.o src/v/io/CMakeFiles/v_io.dir/page.cc.o src/v/io/CMakeFiles/v_io.dir/page_set.cc.o src/v/io/CMakeFiles/v_io.dir/clang-tidy-helper.cc.o src/v/io/CMakeFiles/v_io.dir/logger.cc.o src/v/io/CMakeFiles/v_io.dir/io_queue.cc.o src/v/io/CMakeFiles/v_io.dir/scheduler.cc.o  -Wl,-rpath,/vectorized/lib64  /vectorized/lib64/libseastar.so  /vectorized/lib64/libboost_program_options.so.1.82.0  /vectorized/lib64/libboost_thread.so.1.82.0  /vectorized/lib64/libboost_atomic.so.1.82.0  /vectorized/lib64/libboost_chrono.so.1.82.0  /vectorized/lib64/libboost_date_time.so.1.82.0  /vectorized/lib64/libboost_container.so.1.82.0  /vectorized/lib64/libboost_exception.a  /vectorized/lib64/libcares.so  /vectorized/lib64/libcryptopp.so  /vectorized/lib64/libfmtd.so.8.1.1  -Wl,--as-needed  /vectorized/lib64/liblz4.so  -fsanitize=address  -fsanitize=undefined  -fno-sanitize=vptr  /vectorized/lib64/libabsl_cord.so.2308.0.0  /vectorized/lib64/libabsl_cordz_info.so.2308.0.0  /vectorized/lib64/libabsl_cord_internal.so.2308.0.0  /vectorized/lib64/libabsl_cordz_functions.so.2308.0.0  /vectorized/lib64/libabsl_cordz_handle.so.2308.0.0  /vectorized/lib64/libabsl_crc_cord_state.so.2308.0.0  /vectorized/lib64/libabsl_crc32c.so.2308.0.0  /vectorized/lib64/libabsl_crc_internal.so.2308.0.0  /vectorized/lib64/libabsl_crc_cpu_detect.so.2308.0.0  /vectorized/lib64/libabsl_str_format_internal.so.2308.0.0  /vectorized/lib64/libabsl_raw_hash_set.so.2308.0.0  /vectorized/lib64/libabsl_hash.so.2308.0.0  /vectorized/lib64/libabsl_bad_optional_access.so.2308.0.0  /vectorized/lib64/libabsl_city.so.2308.0.0  /vectorized/lib64/libabsl_bad_variant_access.so.2308.0.0  /vectorized/lib64/libabsl_low_level_hash.so.2308.0.0  /vectorized/lib64/libabsl_hashtablez_sampler.so.2308.0.0  /vectorized/lib64/libabsl_exponential_biased.so.2308.0.0  /vectorized/lib64/libabsl_synchronization.so.2308.0.0  /vectorized/lib64/libabsl_stacktrace.so.2308.0.0  /vectorized/lib64/libabsl_graphcycles_internal.so.2308.0.0  /vectorized/lib64/libabsl_kernel_timeout_internal.so.2308.0.0  /vectorized/lib64/libabsl_time.so.2308.0.0  /vectorized/lib64/libabsl_civil_time.so.2308.0.0  /vectorized/lib64/libabsl_time_zone.so.2308.0.0  /vectorized/lib64/libabsl_symbolize.so.2308.0.0  /vectorized/lib64/libabsl_strings.so.2308.0.0  /vectorized/lib64/libabsl_string_view.so.2308.0.0  /vectorized/lib64/libabsl_throw_delegate.so.2308.0.0  /vectorized/lib64/libabsl_strings_internal.so.2308.0.0  /vectorized/lib64/libabsl_int128.so.2308.0.0  /vectorized/lib64/libabsl_debugging_internal.so.2308.0.0  /vectorized/lib64/libabsl_malloc_internal.so.2308.0.0  /vectorized/lib64/libabsl_demangle_internal.so.2308.0.0  /vectorized/lib64/libabsl_base.so.2308.0.0  /vectorized/lib64/libabsl_raw_logging_internal.so.2308.0.0  /vectorized/lib64/libabsl_log_severity.so.2308.0.0  /vectorized/lib64/libabsl_spinlock_wait.so.2308.0.0  /vectorized/lib64/libroaring.so  -lunordered_dense && :
ld.lld: error: unable to find library -lunordered_dense
```

internal build is failing because `cluster` doesn't depend on `v::container`.

StephanDollberg replied:
Why does it try to link against unordered_dense though, it's a header-only library.

rockwotj replied:
Oh because you put it in deps, it will try and link against it

rockwotj replied:
Maybe trying adding `unordered_dense::unordered_dense` instead?

rockwotj replied:
Oh that will only work for the OSS build. There is no cmake target for this in the superbuild so cmake tries to link against it, I think thats why the :: alias pattern emerged. I am sort of out of my depth here, I am not a CMake wizard

rockwotj replied:
Thoughts: either need a conditional add library for oss or make a dummy target in vtools

StephanDollberg replied:
Thanks for those pointers, so I also have no idea what I am doing here but with: https://github.com/redpanda-data/vtools/compare/stephan/find-unordered-dense?expand=1 

and 

```
diff --git a/src/v/container/CMakeLists.txt b/src/v/container/CMakeLists.txt
index 7b260ee289..c524beffdf 100644
--- a/src/v/container/CMakeLists.txt
+++ b/src/v/container/CMakeLists.txt
@@ -1,7 +1,11 @@
 find_package(Roaring REQUIRED)
+find_package(unordered_dense REQUIRED)
+
 v_cc_library(
   NAME container
-  DEPS Roaring::roaring unordered_dense
+  DEPS
+    Roaring::roaring
+    unordered_dense::unordered_dense
 )

 add_subdirectory(tests)
```

it seems to work. 

@dotnwat is this the right way to go?

We really need to get away from the dual build.

dotnwat replied:
Everything looks ok.

When it's not a header-only library we can usually get away with a little alias hack in the open-source build like this https://github.com/redpanda-data/redpanda/blob/dev/cmake/dependencies.cmake#L135-L137.

But given it's header-only I think [FindUnorderedDense.cmake](https://github.com/redpanda-data/vtools/compare/stephan/find-unordered-dense?expand=1#diff-c270fc1d95547bc979eb41db03cc46f894fb9c10b2df03e29b7a627f70929b15) is the most reasonable thing. And that looks fine.

> We really need to get away from the dual build.

100%

StephanDollberg replied:
Thanks Noah, here it is: https://github.com/redpanda-data/vtools/pull/2594

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/46672#018e6d93-1d0c-40bc-b583-87cf7d32e2fa
<!--12d673f5-d1b3-435a-bc0d-723c18831719-->

rockwotj reviewed (approved):
public build passes: https://github.com/redpanda-data/redpanda/actions/runs/8419599090

rockwotj commented:
@StephanDollberg let's update the advice on the wiki with this: https://redpandadata.atlassian.net/wiki/spaces/CORE/pages/318275653/Memory+Management+in+Redpanda

vbotbuildovich commented:
/backport v23.3.x
